### ISSUE TITLE
add a non-mpi version of teca library

### ISSUE THREAD

**Issue by taobrienlbl** [LBL-EESA/TECA#376]
It would be useful to be able to do `import teca` without requiring MPI calls: particularly when attempting to debug/test python-based pipelines on systems like NERSC, which prohibit using MPI on login nodes.

Ideally, it would be something like:
```python
import teca_nompi

# build TECA pipeline
```

**Comment by burlen** [comment]
It would be easier to have 2 installs. There are some options in the superbuild for instance
```
-DWITHOUT_MPI=ON
```
Tells it not to build any MPI. In that case one would also want to set the install prefix to an alternate location
```
-DCMAKE_INSTALL_PREFIX/some/path/teca_no_mpi
```
then
```
module use /some/path/teca_no_mpi/modulefiles
module load teca
```
to get at the no mpi install. As an alternative one can tell the superbuild to compile MPICH and then one can use MPI on the login node as well.

@taobrienlbl what do you think about the 2 install approach?

**Comment by burlen** [comment]
@taobrienlbl it should be possible to `import teca` from an MPI install of teca when running on a login node. Some of the command line apps do this. For example `teca_convert_table` works fine on the login node.

Things crash when the code also does `from mpi4py import *` since that is what calls `MPI_Init` and Cray makes it abort on the login node.

A work around for the parallel command line apps on the login node might be to use an environment variable `export TECA_NO_MPI=1` (or something like this) which could be tested for before trying to `import mpi4py` or calling `MPI_Init` in the C++ apps.

What do you think?

**Comment by taobrienlbl** [comment]
The two-install approach could work, though I'm somewhat partial to your idea of using an environment variable to control the behavior.  That would allow us to use a single install, which I think would be much simpler from a user point of view.  But if it's difficult to implement, I'm not sure I'd push strongly for this feature.

By the way, I can confirm that simply importing teca does not work on cray:
```bash
$ module load teca
$ python3 -c 'import teca'
[Tue Jun 23 17:40:46 2020] [unknown] Fatal error in PMPI_Init_thread: Other MPI error, error stack:
MPIR_Init_thread(537):
MPID_Init(246).......: channel initialization failed
MPID_Init(647).......:  PMI2 init failed: 1
Aborted
```

It appears that `import teca` calls `from mpi4py import *` at some point, or the SWIG code invokes the C variant of MPI_Init() -- I'd have to look more closely to be sure.  And `teca_convert_table` does not work for me on the login node: same result. Is it possible you're using a build that linked against a non-cray MPICH or MPI?

Thanks thinking through several options.

**Comment by burlen** [comment]
@taobrienlbl This used to work, it is in  fact a bug that crept in. Fortunately I think it can be fixed fairly easily.

**Comment by burlen** [comment]
@taobrienlbl Here is a potential solution that doesn't require changes to TECA. In your script you could do the following.
```Python
import mpi4py
mpi4py.rc(initialize=False, finalize=False)
from teca import *
```

**Comment by taobrienlbl** [comment]
That's good to know about `mpi4py.rc`--thanks! 

Yes, that does work in the particular case that I was testing.  Perhaps we should document that as an option for testing on cray login nodes.  The caveat is that any MPI-enabled TECA code would (I assume) error-out when any MPI function calls are made: so this precludes adding anything like teca_programmable_reduce to a pipeline.

That issue aside, your suggestion points to a potentially easy solution that we could add to TECA; something like the following near the top of `teca.py` or `teca_py.py`:

```python
using_mpi = True
try:
    if os.environ['TECA_NO_MPI'] == 1:
        using_mpi = False
except:
    pass

if not using_mpi:
    import mpi4py
    mpi4py.rc(initialize=False, finalize=False)
```

**Comment by burlen** [comment]
> The caveat is that any MPI-enabled TECA code would (I assume) error-out when any MPI function calls are made:

Only if we are careless  about making MPI calls. Usually we are careful to check if MPI is initialized before making any MPI call. I have found a couple spots in the cf writer where I was careless and I will fix these.

> a potentially easy solution that we could add to TECA; something like the following near the top of teca.py or teca_py.py:

Perhaps it is better to put such logic in the command line apps (that are expected to run in both places). And let users do the same if they'd like to run their scripts on the login nodes. Your proposal may not work because I think we are calling a c function to initialize mpi4py, which I believe is causing the trouble. that was there for a reason (I don't recall why) but it may not be easy to get rid of. Will have to see.

**Comment by burlen** [comment]
The C call `import_mpi4py()` is not ours, but rather part of mpi4py itself injected by mpi4py SWIG code. Unfortunately in addition to calling `MPI_Init` this function does all kinds of low level internal initialization of the mpi4py Python module itself. It is not possible to remove it.

**Comment by burlen** [comment]
#381 implements Travis's suggestion. One caveat is that when MPI_Init is not called and `mpiexec -n N` is used `N` serial copies of the program are invoked. This is never useful. For that reason, I suggest that this feature should be an undocumented(with respect to users) "super user" feature to avoid any confusion. Especially since I have observed a few team members hard coding environment variables in their bashrc, and it causing confusing  problems when mixed with TECA.

**Comment by burlen** [comment]
Resolved in #381
